blockingconcurrentqueue.h

```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+	// instead of copied.
+	// Does not allocate memory (except for one-time implicit producer).
+  // Thread-safe.
+	template<typename It, typename Rep, typename Period>
+  inline void wait_enqueue_bulk_timed(It itemFirst, size_t max, std::chrono::duration<Rep, Period> const& timeout) {
+    return wait_enqueue_timed<It&>(itemFirst, max, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
+  }
+
+  // Enqueues several items using an explicit producer token.
+  // Blocks if necessary before (and between) inserting items if not
+  // enough space is available for all the items.
+  // May enqueue less than count elements, if the timeout expires before
+  // all elements could be added.
+  // The number of elements added is returned.
+	// Note: Use std::make_move_iterator if the elements should be moved
+	// instead of copied.
+	// Does not allocate memory (except for one-time implicit producer).
+  // Thread-safe.
+	template<typename It>
+  inline void wait_enqueue_bulk_timed(producer_token_t const& token, It itemFirst, size_t max, std::int64_t timeout_usecs)
+  {
+    std::int64_t t0 = details::getutime(), tnow = t0;
+    size_t count = 0;
+    while (count < max && (tnow - t0) < timeout_usecs) {
+		  size_t gained = (size_t)sema_enqueue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)(max - count), tnow - t0);
+      while (gained > 0 && !inner.try_enqueue_bulk(token, std::forward<It&>(itemFirst), gained)) {
+        continue;
+      }
+      count += gained;
+      tnow = details::getutime();
+    }
+  }
+
+  // Enqueues several items using an explicit producer token.
+  // Blocks if necessary before (and between) inserting items if not
+  // enough space is available for all the items.
+  // May enqueue less than count elements, if the timeout expires before
+  // all elements could be added.
+  // The number of elements added is returned.
+	// Note: Use std::make_move_iterator if the elements should be moved
+	// instead of copied.
+	// Does not allocate memory (except for one-time implicit producer).
+  // Thread-safe.
+	template<typename It, typename Rep, typename Period>
+  inline void wait_enqueue_bulk_timed(producer_token_t const& token, It itemFirst, size_t max, std::chrono::duration<Rep, Period> const& timeout) {
+    return wait_enqueue_timed<It&>(token, itemFirst, max, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
+  }
+
 	// Attempts to dequeue from the queue.
 	// Returns false if all producer streams appeared empty at the time they
 	// were checked (so, the queue is likely but not guaranteed to be empty).
+  // Also returns false if the queue is finished (closed and drained of all remaining elements).
 	// Never allocates. Thread-safe.
 	template<typename U>
 	inline bool try_dequeue(U& item)
 	{
-		if (sema->tryWait()) {
+		if (sema_dequeue->tryWait()) {
 			while (!inner.try_dequeue(item)) {
 				continue;
 			}
+			sema_enqueue->signal();
```

Comment:
Because elements are stored in blocks, dequeueing an item doesn't free up a slot for enqueueing right away. Not sure how to reconcile this.

Reply:
Well, the enqueue ops will spin so I think it would be work, but it's not very efficient :/

Reply:
They could end up spinning indefinitely if only one item is ever dequeued :P